src/datasets/arrow_dataset.py

```diff
@@ -2809,10 +2813,58 @@ def map(
                 f"num_proc must be <= {len(self)}. Reducing num_proc to {num_proc} for dataset of size {len(self)}."
             )
 
-        disable_tqdm = not logging.is_progress_bar_enabled()
+        def load_processed_shard_from_cache(shard_kwargs):
+            """Load a processed shard from cache if it exists, otherwise throw an error."""
+            shard = shard_kwargs["shard"]
+            if shard_kwargs["new_fingerprint"] is None:
+                transform = format_transform_for_fingerprint(Dataset._map_single)
+                kwargs_for_fingerprint = format_kwargs_for_fingerprint(
+                    Dataset._map_single, (), shard_kwargs, ignore_kwargs=["cache_file_name"]
+                )
+                kwargs_for_fingerprint["fingerprint_name"] = "new_fingerprint"
+                shard_kwargs["new_fingerprint"] = update_fingerprint(
+                    shard._fingerprint, transform, kwargs_for_fingerprint
+                )
+            else:
+                validate_fingerprint(shard_kwargs["new_fingerprint"])
+
+            # Check if we've already cached this computation (indexed by a hash)
+            if shard.cache_files:
+                if shard_kwargs["cache_file_name"] is None:
+                    # we create a unique hash from the function,
+                    # current dataset file and the mapping args
+                    shard_kwargs["cache_file_name"] = shard._get_cache_file_path(shard_kwargs["new_fingerprint"])
+                if os.path.exists(shard_kwargs["cache_file_name"]) and load_from_cache_file:
+                    logger.warning(f"Loading cached processed dataset at {shard_kwargs['cache_file_name']}")
+                    info = shard.info.copy()
+                    info.features = features
+                    info.task_templates = None
+                    return Dataset.from_file(shard_kwargs["cache_file_name"], info=info, split=shard.split)
+            raise NonExistentDatasetError
+
+        num_shards = num_proc if num_proc is not None else 1
+        if batched and drop_last_batch:
+            pbar_total = len(self) // num_shards // batch_size * num_shards * batch_size
+        else:
+            pbar_total = len(self)
+
+        # Inititialize the progress bar later inside the loop to avoid the progress bar being displayed when there is no processing
+        pbar = None
+
+        def init_pbar():
+            nonlocal pbar
+            pbar = logging.tqdm(
+                disable=not logging.is_progress_bar_enabled(),
+                unit=" examples",
+                total=pbar_total,
+                leave=False,
+                desc=f"Processing the dataset ({shards_done}/{num_shards} shards)" if desc is None else desc,
+            )
```

Comment:
wouldn't it be simpler to instantiate the pbar just before the `for` loop instead of inside ? this way no need for `init_pbar()`